---
title: "Correcting batch effects"
output: html_notebook
---

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("sva", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
# library(sva)
```
The 'sva' package can be used to apply the ComBat method to correct batch effects in micro array and sequencing data. I will first do the tutorial to learn how to use the package.
```{r}
# biocLite("bladderbatch")
# biocLite("limma")
library(sva)
library(limma)
library(bladderbatch)
data(bladderdata)
```
```{r}
pheno = pData(bladderEset)
edata = exprs(bladderEset)
mod = model.matrix(~as.factor(cancer), data=pheno)
mod0 = model.matrix(~1, data = pheno)
```
```{r}
n.sv <- num.sv(edata, mod, method="leek")
svobj <- sva(edata, mod, mod0, n.sv= n.sv)
```
```{r}
pValues <- f.pvalue(edata, mod, mod0)
qValues <- p.adjust(pValues, method="BH")
hist(pValues, cex = 0.1)
```
Now, correcting the P- and Q-values for surrogate variables:
```{r}
modSv <- cbind(mod, svobj$sv)
mod0Sv <- cbind(mod0, svobj$sv)
pValuesSv <- f.pvalue(edata, modSv, mod0Sv)
qValuesSv <- p.adjust(pValuesSv, method="BH")
hist(pValuesSv, cex = 0.1)
hist(qValuesSv, cex = 0.1)
```
Using the *limma* package to adjust for surrogate variables:
```{r}
fit <- lmFit(edata, modSv)
```
Finally, using the *ComBat* function to adjust for known batches (this is what I actually need):
```{r}
batch <- pheno$batch
modcombat <- model.matrix(~1, data = pheno)
combat_edata <- ComBat(dat=edata, batch=batch, mod=modcombat, par.prior = TRUE)
```
This returns an expression matrix that has been adjusted for batch. Significance analysis:
```{r}
pValuesCombat <- f.pvalue(combat_edata, mod, mod0)
qValuesCombat <- p.adjust(pValuesCombat, method="BH")
plot(density(pValuesCombat))
```

Now, applying this technique to the RTCGA data:
```{r}
diseases <- c("BRCA", "GBM", "OV")
omics <- c("mRNA", "miRNASeq", "methylation")
for(d in diseases){
  assign(
    sprintf("%s_pheno", d),
    read.table(
      sprintf("%s_clinical.txt_with_survival_data.txt", d)
    ),
    envir = .GlobalEnv
  )
  for(o in omics){
    assign(sprintf("%s_%s", d, o), 
           read.table(sprintf("%s_%s_with_survival_data.txt", d, o)),
    envir = .GlobalEnv)
    assign(sprintf("%s_%s_patients", d, o),
           read.table(sprintf("%s_%s_patients.txt_with_survival_data.txt", d, o)),
    envir = .GlobalEnv)
  }
}
```
The problem is I need to supply batch ID's for all the samples, and I'm not sure what I can use as batch ID. I thought about using the MBatch package because it is specified for TCGA data, but also for this you need to supply a batch ID, and it is not documented well (at least the installation isn't).

https://confluence.broadinstitute.org/display/GDAC/FAQ: the plate identifier that is used in the TCGA barcode can effectively be used as batch ID. Therefore I need to add a column "batch_ID" to the different omic dataframes.

20-02-2017:

* Extract batch ID from patient barcode in AddSurvDataToOmics()
    + Done:
```{r}
plot(as.factor(BRCA_mRNA$batch_id))
```
* Write a function to correct batch effects
* Correct the data:
```{r}
test_mat <- t(as.matrix(BRCA_mRNA[]))
test_pheno <- 
CorrectBatch()
```
* Since the ComBat() function wants a matrix with samples on the columns and genes on the rows, it is probably a better idea to implement this in the AddSurvDataToOmics() function
    + potential problem 1: there are two variables of interest for prediction
    + potential problem 2: both are continuous, not categorical
    + **Read more about the ComBat() function**
    + I was confused... The batch ID has to do with the samples of course, not with the patients
* Save the corrected data (with different file names from the uncorrected data)

21-2-2017:

* I need:
    + matrix *genes x samples* with omic data
    + vector *samples x 1* with patient_id's for a certain omic data source (may contain duplicates)
    + dataframe *patients x 4* that contains batch_id and clinical data for all patients from a particular disease
* Test new AddSurvivalDataToOmics() function:
```{r}
# AddSurvDataToOmics("BRCA", "mRNA")
ind <- 1:10
mat_test <- BRCA_mRNA[ind,ind]
pat_test <- BRCA_mRNA_patients[ind,]
phe_test <- BRCA_pheno[as.character(pat_test$patient_id),]
test_pheno_input <- data.frame(cbind(pat_test, BRCA_pheno[as.character(pat_test$patient_id),]))
```

```{r}
cor <- CorrectBatch(BRCA_mRNA, BRCA_mRNA_patients, BRCA_pheno)
```

22-02-2017

Of course, now we have missing values because either days_to_death or days_to_last_followup is NA. I need to read more about censoring in survival analysis.

Random thought: days_to_death and days_to_followup are counted from the day of diagnosis, right? So what if some people just went to the doctor earlier? This skews the results...

Second thought: would it maybe be a good idea to discretize the survival data in short and long survival before correcting the batch effects?

```{r}
# pc_BRCA_mRNA <- prcomp(BRCA_mRNA[!is.na(BRCA_mRNA)])
# library(ggplot2)
# Y.complete <- rbind(Y1, Y2)
# phe.complete <- rbind(phe1,phe2)
# pc <- prcomp(Y.complete)
pc <- prcomp(BRCA_mRNA)
```
There are missing values in this dataset...
```{r}
diseases <- c("BRCA", "GBM", "OV")
omics <- c("mRNA", "miRNASeq", "CNV", "methylation")
a <- paste("RTCGA.", c("clinical", omics), sep = "")
b <- paste(a, ".20160128", sep = "")
pckgs <- c(a,b)

library("RTCGA", lib.loc = "../R/x86_64-pc-linux-gnu-library/3.3")
for(p in pckgs){
  require(p, character.only = TRUE, lib = "../R/x86_64-pc-linux-gnu-library/3.3")
}

```
```{r}
pc <- prcomp(BRCA_mRNA)
pc.pred <- predict(pc,newdata = BRCA_mRNA)
p <- ggplot(as.data.frame(pc.pred), aes(x=pc.pred[,1], y= pc.pred[,2],
colour= scale_color_gradient(BRCA_pheno$patient.days_to_death[BRCA_mRNA_patients]))) + ggtitle("Batch Effects before
Transformation") + labs(y = "PC1", x = "PC2", colour = "Classes") +
geom_point(shape=19)
plot(p)
```
**Intermezzo**
New problem: for CNV, methylation and miRNASeq I don't want a patients x genes matrix (well, in the end I do but not for now) and I also don't need to correct batch effects for this (is this correct??).

So, change the addsurvdata function again to account for this.
*Done*
Also changed the intersection function to be able to work with the new way of reading in the data.

Okay, so there are missing values... This is something that should be taken care of during the reading in of the data...
```{r}
missing <- is.na(BRCA_mRNA)
length(missing)
sum(missing)
```

1,695 missing values on 10,510,260 total. I want to know which rows and columns contains these missing values, but I don't know how to get this information.
```{r}
sum(is.na(GBM_mRNA))
```

```{r}
sum(is.na(OV_mRNA))
```

Ashar mentioned there is a package to impute missing data, specifically for micro array data: impute.
```{r}
biocLite("impute")
library(impute)
```
```{r}
for(d in diseases){
  assign(
    sprintf("%s_mRNA_imputed", d),
    impute.knn(get(sprintf("%s_mRNA", d))),
    envir = .GlobalEnv
  )
}
```

```{r}
# library(ggplot2)
# Y.complete <- rbind(Y1, Y2)
# phe.complete <- rbind(phe1,phe2)
# pc <- prcomp(BRCA_mRNA_imputed$data)
# pc.pred <- predict(pc,newdata = BRCA_mRNA_imputed$data)
# p <- ggplot(as.data.frame(pc.pred), aes(x=pc.pred[,1], y= pc.pred[,2]))
# + ggtitle("Batch Effects before Transformation") 
# + labs(y = "PC1", x = "PC2", colour = "Classes") + geom_point(shape=19)
p <- ggplot(as.data.frame(pc.pred), aes(x=pc.pred[,1], y= pc.pred[,2])) + ggtitle("Batch Effects before Transformation") + labs(y = "PC1", x = "PC2", colour = "Classes") + geom_point(shape=19)
p
```

Next problem: the survival data as I'm using it now has NA's. It might be a better idea to make one column with days_to_death and one column saying if the particular patient was censored (boolean values).
```{r}
test <- ifelse(is.na(BRCA_pheno$patient.days_to_death), BRCA_pheno$patient.days_to_last_followup, BRCA_pheno$patient.days_to_death)
```
This works, but now I don't know anymore which patients are censored.
```{r}
days <- ifelse(is.na(BRCA_pheno$patient.days_to_death), 
               BRCA_pheno$patient.days_to_last_followup, 
               BRCA_pheno$patient.days_to_death)
censored <- ifelse(is.na(BRCA_pheno$patient.days_to_death),
                   TRUE,
                   FALSE
                   )
new_pheno <- cbind(days, censored)
```
```{r}
# pc <- prcomp(t(BRCA_mRNA_imputed$data))
# pc.pred <- predict(pc,newdata = t(BRCA_mRNA_imputed$data))
p <- ggplot(as.data.frame(pc.pred), aes(x=pc.pred[,1], y= pc.pred[,2], colour = as.numeric(new_pheno[BRCA_mRNA_patients$patient_id,"days"]))) + ggtitle("Batch Effects before Transformation") + labs(y = "PC1", x = "PC2", colour = "Classes") + geom_point(shape=19)
p
```

```{r}
b <- ggplot(as.data.frame(pc.pred), aes(x=pc.pred[,1], y= pc.pred[,2], colour = as.factor(BRCA_mRNA_patients$batch_id))) + ggtitle("Batch Effects before Transformation") + labs(y = "PC1", x = "PC2", colour = "Classes") + geom_point(shape=19)
b
```

The batch effect doesn't seem to be really strong, it is probably best to not correct because the more the data is changed the more we could bias the outcome. But, check for the other diseases and for miRNASeq as well.

* Write a function that extracts only 'read_counts' or 'reads_per_million_mirna_mapped' and puts it in a matrix
* Write a function that makes a plot like the one seen above and saves it in a variable
* Call this function for each mRNA or miRNA - disease combination

24-02-2017

Group meeting:

* Definitely use ComBat to correct the batch effects: they might not always be visible directly in a PCA plot, but could still have an effect
* And still need to check the other data sets.
* Download micro array data instead of sequencing data.
  - I can not find miRNA micro array data on TCGA (not even on the website itself)
  
28-02-2017

```{r}
for(d in c("BRCA", "OV")){
  print(d)
  assign(sprintf("%s_mRNA_corrected", d), CorrectBatch(
    get(sprintf("%s_mRNA_imputed", d))$data,
    get(sprintf("%s_mRNA_patients", d)),
    get(sprintf("%s_pheno", d))
  ))
}
```
GBM data is all from one batch, so doesn't need to be corrected.

```{r}
library("VennDiagram", lib.loc = "../R/x86_64-pc-linux-gnu-library/3.3")
venn <- venn.diagram(list(local = GBM_mRNA_patients$patient_id, new = rownames(GBM_pheno)), filename = NULL)
grid.newpage(recording = FALSE)
grid.draw(venn, recording = FALSE)
```

```{r}
library("VennDiagram", lib.loc = "../R/x86_64-pc-linux-gnu-library/3.3")
venn <- venn.diagram(list(local = OV_mRNA_patients$patient_id, new = rownames(OV_pheno)), filename = NULL)
grid.newpage(recording = FALSE)
grid.draw(venn, recording = FALSE)
```

Problem: there is no overlap in GBM mRNA data and GBM clinical data. Aha, I found the reason: in reading in the data, there should be one choice made for 'new' or 'old' data, not change it per omic data.

```{r}
source("https://bioconductor.org/biocLite.R")
diseases <- c(TRUE, FALSE, TRUE)
names(diseases) <- c("BRCA", "GBM", "OV")
omics <- c("mRNA", "miRNASeq", "CNV", "methylation")
a <- paste("RTCGA.", c("clinical", omics), sep = "")
b <- paste(a, ".20160128", sep = "")
pckgs <- c(a,b)

library("RTCGA", lib.loc = "../R/x86_64-pc-linux-gnu-library/3.3")
for(p in pckgs){
  require(p, character.only = TRUE, lib = "../R/x86_64-pc-linux-gnu-library/3.3")
}

source("RTCGA_functions.R")

AddSurvDataToOmics(diseases, omics, save = TRUE)

# calculate the intersection (which patients have data for all omics types)
for(dis in names(diseases)){
  assign(sprintf("%s_inter", dis), FindIntersection(omics, dis, save = TRUE))
}

# biocLite("impute")
library(impute)

for(d in c("BRCA", "OV")){
  assign(
    sprintf("%s_mRNA_imputed", d),
    impute.knn(get(sprintf("%s_mRNA", d))),
    envir = .GlobalEnv
  )
}

library("sva")
for(d in c("BRCA", "OV")){
  print(d)
  assign(sprintf("%s_mRNA_corrected", d), CorrectBatch(
    get(sprintf("%s_mRNA_imputed", d))$data,
    get(sprintf("%s_mRNA_patients", d)),
    get(sprintf("%s_pheno", d))
  ))
}
```