---
title: "Adding clinical data to omics data"
output: html_notebook
---
I'm installing directly from Github instead of using Bioconductor, because:
- I'm using Rstudio on a server, and I could not manage to properly install bioconductor there
- The (stable) Github version is the most recent
```{r}
# This works now, didn't work before with the old version of Rstudio...
source("http://bioconductor.org/biocLite.R")
if (!require(devtools)) {
    install.packages("devtools", lib = "../r-packages")
    require(devtools)
}
```

```{r}
a <- paste("RTCGA.", c("clinical", omics), sep = "")
b <- paste(a, ".20160128", sep = "")
pckgs <- c(a,b)

# setRepositories(ind = c(1:6, 8))
# biocLite("RTCGA", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
```
That works, but this doesn't:
```{r}
for(p in pckgs){
  # install_github(p, build_vignettes = TRUE, lib = "../r-packages")
  # biocLite(p)#, lib = "../r-packages")
  # installTCGA(p)
  library(p, character.only = TRUE, lib.loc = "../R/x86_64-pc-linux-gnu-library/3.3")
  }
```
The Rstudio on the server is updated to 3.3.2. I am now trying to install the necessary packages by using the command line R instead of the Rstudio server (after manually removing all installed packages using the command line):
```{r}
setRepositories(ind = c(1:6, 8))
source("http://bioconductor.org/biocLite.R")
biocLite("RTCGA")
library("RTCGA")
installTCGA()
```

Nope. On command line R (R \-\-vanilla):

```{r}
# I can't use https on the R-server
options(repos='http://cran.rstudio.com/')
install.packages("devtools")
library("devtools")#, lib = "../R/x86_64-pc-linux-gnu-library/3.3")
install_github("RTCGA/RTCGA", build_vignettes = TRUE)

diseases <- c("BRCA", "GBM", "OV")
omics <- c("mRNA", "miRNASeq", "CNV", "methylation")
a <- paste("RTCGA.", c("clinical", omics), sep = "")
b <- paste(a, ".20160128", sep = "")
pckgs <- c(a,b)

for(p in pckgs){
  install_github(paste("RTCGA/", p, sep = ""), build_vignettes = TRUE, force = TRUE)
}
```
And then, on the server:
```{r}
library("RTCGA", lib.loc = "../R/x86_64-pc-linux-gnu-library/3.3")
for(p in pckgs){
  library(p, character.only = TRUE, lib = "../R/x86_64-pc-linux-gnu-library/3.3")
}
```
It worked! I can now use the most recent data from the RTCGA packages.

```{r}
source("RTCGA_functions.R")

AddSurvDataToOmics(diseases, omics, save = TRUE)

# calculate the intersection (which patients have data for all omics types)
for(dis in diseases){
  assign(sprintf("%s_inter", dis), FindIntersection(omics, dis, save = TRUE))
}
```
14-02-2017:

* Since it is not clear what I should use as class label for the PRAD patients, and most samples in the dataset I have downloaded are healthy, Ashar and I have decided to not look at PRAD for now.
* Check CNV data
    + CNV data is not so straightforward to analyze, it needs to be converted to a patients x genes matrix in some way
    + **Read the original paper**
* Check on TCGA website if the GBM datasets are correctly loaded using RTCGA, since the number of patients seems low
    + this is happening correctly
    + **Read original paper that came out with the GBM patients.**
* Check out the local data (abidata)
    + The lengths of the intersections are as follows: 172 patients for BRCA, 274 patients for GBM, 435 for OV
* Check intersections of these datasets.
* Think about adding m(i)rRNA instead of m(i)RNASeq (Seq has a Poisson distribution instead of normal distribution). It is probably better to add the micro-array data (so not Seq) because it is easier to model.
    + The local data is micro array, on TCGA only Seq data is available for miRNA
* **Think about using SQL to make faster connections between clinical data and omics data for a specific patient**
* Add methylation data.
```{r}
# First install methylation packages:
library("RTCGA", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
library("RTCGA.clinical", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
library("RTCGA.clinical.20160128", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
library("RTCGA.methylation", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
library("RTCGA.methylation.20160128", lib = "../R/x86_64-pc-linux-gnu-library/3.3")
```
```{r}
# Then process the methylation data:
AddSurvDataToOmics(diseases, omics, save = TRUE)
for(dis in diseases){
  assign(sprintf("%s_inter", dis), FindIntersection(c(omics, "methylation"), dis, save = FALSE))
}
```


