
---
title: "R Notebook"
output: html_notebook
---

```{r}
# biocLite("KEGGgraph")
# biocLite("UniProt.ws")
library(KEGGgraph)
library(UniProt.ws)
```

KEGGgraph is probably great for using the topological info from pathways, but for now I just want genelists.

```{r}
# biocLite("org.Hs.eg.db")
library(org.Hs.eg.db)
kegg <- org.Hs.egPATH2EG
mapped <- mappedkeys(kegg)
kegg2 <- as.list(kegg[mapped])
```
Oh joy... I need to convert identifiers. =D

```{r}
uniprots <- Rkeys(org.Hs.egUNIPROT)
# genes_BRCA_mRNA <- colnames(BRCA_mRNA)
# genes_BRCA_mRNA[1:4]
# genes <- UniProt.ws(genes_BRCA_mRNA)
select(org.Hs.eg.db, uniprots[1:4], "ENTREZID", "UNIPROT")
```
```{r}
# biocLite("GenomicFeatures")
library(GenomicFeatures)
# biocLite("mygene")
library(mygene)
```
```{r}
## Bimap interface:
x <- org.Hs.egGENENAME
# Get the gene names that are mapped to an entrez gene identifier
mapped_genes <- mappedkeys(x)
# Convert to a list
xx <- as.list(x[mapped_genes])
```

```{r}
gene_lookup <- cbind(org.Hs.eg.db, getGene(org.Hs.eg.db)$name)
```
```{r}
# biocLite("biomaRt")
library(biomaRt)
```

```{r}
ensembl <- useMart("ensembl",dataset="hsapiens_gene_ensembl")
 
#create a filter for all assembled human chromosomes
my_chr <- c(1:22, 'M', 'X', 'Y')
 
#listAttributes shows all attributes
attributes <- listAttributes(ensembl)
 
#find entrez attribute name
grep(pattern="entrez", x=attributes$description, ignore.case=T)
#[1] 62
attributes[62,]
# entrezgene

grep(pattern="hgnc symbol", x=attributes$description, ignore.case=T)
attributes[71,]
# hgnc_symbol
```
```{r}
entrez_hgnc <- getBM(attributes = c('entrezgene', "hgnc_symbol"),
                        filters = 'chromosome_name',
                        values = my_chr,
                        mart = ensembl
                        )
entrez_hgnc <- entrez_hgnc[complete.cases(entrez_hgnc),]
entrez_hgnc <- entrez_hgnc[entrez_hgnc$entrezgene != "" & entrez_hgnc$hgnc_symbol != "",]
dupli_hgnc <- duplicated(entrez_hgnc$hgnc_symbol)
entrez_hgnc <- entrez_hgnc[!dupli_hgnc,]

rownames(entrez_hgnc) <- entrez_hgnc$hgnc_symbol

```
The test:
```{r}
entrez_BRCA_mRNA <- entrez_hgnc[rownames(BRCA_mRNA), "entrezgene"]
names(entrez_BRCA_mRNA) <- rownames(BRCA_mRNA)
pw1_genes <- entrez_BRCA_mRNA[as.numeric(kegg2[[1]])]
# pathway1_BRCA_mRNA <- apply(BRCA_mRNA[names(pw1_genes),], 1, mean)
```

```{r}
PathwaysMrna <- function(i){
  ind <- rownames(BRCA_mRNA) %in% names(entrez_BRCA_mRNA[as.numeric(kegg2[[i]])])
  if(sum(ind) == 0){return(NULL)}
  if(sum(ind) == 1){return(BRCA_mRNA[ind,])}
  return(apply(BRCA_mRNA[ind,], 2, mean))
}
vPathwaysMrna <- Vectorize(PathwaysMrna)
```
Rewrite the above function to return just one pathway, and cbind the results
```{r}
test4pathways <- vPathwaysMrna(1:4)
colnames(test4pathways) <- names(kegg2[1:4])
```

A few thoughts:

* If I'm collapsing the data on pathways, what is the effect of some data sources having data for a particular gene in a particular pathway and others not having data for that particular gene? Would it be best to remove that gene altogether or not?

13-03-2017
```{r}
# objname <- sprintf("entrez_%s_mRNA", d)
# print(objname)
# data <- get(sprintf("%s_mRNA", d))
# assign(objname, entrez_hgnc[rownames(data), "entrezgene"])
# obj <- get(objname)
# names(obj) <- rownames(data)
# all these objects are identical!
entrez_mRNA <- entrez_hgnc[rownames(BRCA_mRNA), "entrezgene"]
names(entrez_mRNA) <- rownames(BRCA_mRNA)
```

```{r}
# BRCA_mRNA <- BRCA_mRNA_corrected
# OV_mRNA <- OV_mRNA_corrected
save(BRCA_mRNA, file = "/home/abidata/Joanna/BRCA/mRNA/BRCA_mRNA.rda")
save(BRCA_mRNA_patients, file = "/home/abidata/Joanna/BRCA/mRNA/BRCA_mRNA_patients.rda")
save(OV_mRNA, file = "/home/abidata/Joanna/OV/mRNA/OV_mRNA.rda")
save(OV_mRNA_patients, file = "/home/abidata/Joanna/OV/mRNA/OV _mRNA_patients.rda")
save(GBM_mRNA, file = "/home/abidata/Joanna/GBM/mRNA/GBM_mRNA.rda")
save(GBM_mRNA_patients, file = "/home/abidata/Joanna/GBM/mRNA/GBM _mRNA_patients.rda")
```


```{r}
for(d in names(diseases)){
  # objname <- sprintf("%s_mRNA_pathways", d)
  # print(objname)
  print(d)
  obj <- PathwaysMrna(1:229, d)
  # assign(objname, PathwaysMrna(1:229, d))
  colnames(obj) <- names(kegg2)
  assign(sprintf("%s_mRNA_pathways", d), obj)
}
```

05-04-2017

The function for average per pathway was adjusted to work on all omics types instead of only mRNA:

```{r}
# BRCA_CNV <- BRCA_CNV[,c(-1,-2)]
# BRCA_CNV <- cbind(BRCA_CNV)
# BRCA_CNV <- matrix(unlist(BRCA_CNV), nrow = nrow(BRCA_CNV), dimnames = list(rownames(BRCA_CNV), colnames(BRCA_CNV)))
for(d in names(diseases)[1]){
  print(d)
  for(o in omics[3]){
    print(o)
    if(exists(sprintf("%s_%s", d, o))){
      # if(o == "CNV"){
      #   obj <- RUCN(get(sprintf("%s_CNV", d)), get(sprintf("%s_CNV_patients", d)))
      # }
      obj <- AveragePerPathway(1:229, d, o)
      # colnames(obj) <- names(kegg2)
      assign(sprintf("%s_%s_avg", d, o), obj)
    }
  }
}
```

It is not a good idea to use the average operator for CNV data, since these can have negative and positive values. Read more about this, but it might be a good idea to use the fraction of changed genes in a pathway as pathway activity. Another approach might be to undo the -2 transformation to get to to the original data where 0 doesn't mean a copy number of 2 but an actual copy number of 0.

Try this:

```{r}
for(d in names(diseases)){
  objname <- sprintf("%s_CNV", d)
  obj <- get(objname)
  assign(objname, obj+2)
}
```

Well, that was easy. But of course, transforming the data like this does not change the distribution. Now up- and down-regulation will still be canceled out. So back to the idea for fraction of changed or unchanged genes per pathway. If none of the genes in a certain pathway have a changed copy number, I want the ratio to be zero. So, if all the genes have a CNV of zero, I want the result to be zero. If 1/10th of the genes has a CNV of -0.5 and the rest is unchanged, I want the result to be 0.05. If 1/10th of the genes has a CNV of +0.5, I also want the result to be 0.05.

```{r}
# RUCN is Ratio of Unchanged Copy Number
RUCN <- function(data, patients){
  genes <- entrez_hgnc[rownames(data), "entrezgene"]
  res <- do.call(rbind, 
                 lapply(
                   kegg2,
                   function(pw_genes){
                     overlap_entrez <- genes[genes %in% pw_genes]
                     if(length(overlap_entrez) == 0){return(NULL)}
                     overlap_hgnc <- entrez_hgnc[entrez_hgnc$entrezgene == overlap_entrez, 'hgnc_symbol']
                     pw_data <- data[overlap_hgnc,]
                     if(length(overlap_entrez) == 1){return(abs(pw_data)/length(pw_genes))}
                     # do things here
                     names(pw_ratio) <- NULL
                     return(pw_ratio)
                   }
                 )
  )
  colnames(res) <- rownames(patients)
  return(res)
}
```
06-04-2017

Now I've got the situation for an overlap of 0, and an overlap of 1. For all the other situations, I need to think how to do this.

```{r}
test_data <- BRCA_CNV[c(1, 400, 900, 1200),1:10]
test_data
```

```{r}
nr <- nrow(test_data)
res <- apply(test_data, 2, function(pw_genes){return(sum(abs(pw_genes))/nr)})
res
```
```{r}
# RUCN is Ratio of Unchanged Copy Number
RUCN <- function(data, patients){
  genes <- HgncToEntrez(rownames(data))
  res <- do.call(rbind, 
                 lapply(
                   kegg2,
                   function(pw_genes){
                     overlap_entrez <- genes[genes %in% pw_genes]
                     if(length(overlap_entrez) == 0){return(NULL)}
                     overlap_hgnc <- EntrezToHgnc(overlap_entrez)
                     pw_data <- data[overlap_hgnc,]
                     if(length(overlap_entrez) == 1){return(abs(pw_data)/length(pw_genes))}
                     nr <- nrow(pw_data)
                     pw_ratio <- apply(
                       pw_data, 2, function(CNV_pw){
                         return(sum(abs(CNV_pw))/nr)
                       }
                     )
                     names(pw_ratio) <- NULL
                     return(pw_ratio)
                   }
                 )
  )
  colnames(res) <- rownames(patients)
  return(res)
}
```
Now try it out:
```{r}
test_RUCN <- RUCN(BRCA_CNV, BRCA_CNV_patients)
```
I'm getting NA's for some pathways. Let's check them:
```{r}
test_data <- BRCA_CNV[,1:2]
test_data
```
```{r}
nr <- nrow(test_data)
res <- apply(test_data[EntrezToHgnc(kegg2[[2]]),], 2, function(pw_genes){return(sum(abs(pw_genes))/nr)})
res
```
It's fixed now, this function works and is put in RTCGA_functions.R.
```{r}
for(d in names(diseases)){
  obj <- get(sprintf("%s_CNV", d))
  pat <- get(sprintf("%s_CNV_patients", d))
  assign(sprintf("%s_CNV_pwratio", d), RUCN(obj, pat))
}
```

I realized I can very easily rewrite the average function to use absolute values, so it will also work for CNV data. Now I also need to look at the first pc calculation for CNV. That's done, only needed to change the values into absolute values.

NB: PCA and overall mean per pathway are unsupervised techniques to get a pathway activity. Average of 50% genes that are most significantly associated with the clinical outcome is a supervised technique, so the determination of the p-values (in the case of the Cox proportional hazards model) should only be done on a training set. Idea: determine one list of p-values per datatype and disease combination, then when working on pathway level one can take a look at that list. Should the same patients be used as training set for all omics data types? I would say yes.

So before applying this method, I need to choose a training set of patients per disease. These training patients should be a subset of the intersection of patients that occur in every omics data type. Might be good to make sure the class balance in the training set is the same as the balance in the overall dataset (or rather, the overall intersection of patients).