---
title: "First principal component of pathway mappable genes"
output: html_notebook
---

I'm not sure what to do here:
```{r}
BRCA_mRNA <- BRCA_mRNA$data
test <- prcomp(t(BRCA_mRNA)[,])
pc.pred <- predict(test, newdata = t(BRCA_mRNA)[,])
plot(pc.pred[,1], cex = 0.1, col = ifelse(as.numeric(BRCA_pheno$days[BRCA_mRNA_patients$patient_id]) < 300, "red", "green"))
```
Test with one pathway:
```{r}
pw1_genes <- entrez_hgnc[entrez_hgnc$entrezgene %in% kegg2[[1]], "hgnc_symbol"]
pw1 <- BRCA_mRNA[rownames(BRCA_mRNA) %in% pw1_genes,]

prcom <- prcomp(t(pw1))
pc.pred <- predict(prcom, newdata = t(pw1))

plot(pc.pred[,1], cex = 0.1, col = ifelse(as.numeric(BRCA_pheno$days[BRCA_mRNA_patients$patient_id]) < 300, "red", "green"))
```
```{r}
ind2 <- BRCA_mRNA[rownames(BRCA_mRNA) %in% entrez_hgnc[entrez_hgnc$entrezgene %in% kegg2[[2]], "hgnc_symbol"],]
pw2 <- predict(prcomp(t(ind2)), newdata = t(ind2))[,1]

test <- lapply(
  kegg2,
  function(pw_genes){
    ind2 <- BRCA_mRNA[rownames(BRCA_mRNA) %in% entrez_hgnc[entrez_hgnc$entrezgene %in% pw_genes, "hgnc_symbol"],]
    pw2 <- predict(prcomp(t(ind2)), newdata = t(ind2))[,1]
    names(pw2) <- NULL
    return(pw2)
  }
)

test3 <- lapply(
  test,
  function(x){
    return(unlist(x))
  }
)

test4 <- cbind(test)

test2 <- as.matrix(unlist(test), nrow = 229)

test4 <- do.call(rbind, test)
colnames(test4) <- rownames(BRCA_mRNA_patients)

BRCA_mRNA_pw_pc1 <- test4
```

```{r}
FirstPCPerPathway <- function(data, patients, lookup = entrez_hgnc, pwlist = kegg2){
  res <- do.call(rbind, 
                 lapply(
                   pwlist,
                   function(pw_genes){
                     ind2 <- data[
                       rownames(data) %in% lookup[
                         lookup$entrezgene %in% pw_genes, "hgnc_symbol"
                         ],
                       ]
                     pw2 <- predict(prcomp(t(ind2)), newdata = t(ind2))[,1]
                     names(pw2) <- NULL
                     return(pw2)
                   }
                 )
  )
  colnames(res) <- rownames(patients)
  return(res)
}
```

```{r}
BRCA_mRNA_pc1 <- FirstPCPerPathway(BRCA_mRNA_corrected, BRCA_mRNA_patients)
```

```{r}
# rownames(BRCA_CNV) <- BRCA_CNV[,1]
# BRCA_CNV <- BRCA_CNV[,c(-1,-2,-3)]
# BRCA_CNV_patients <- BRCA_CNV_patients[c(-1,-2,-3),]
ind <- seq(2, (nrow(BRCA_miRNASeq)-1), by = 3)
test <- BRCA_miRNASeq[ind,]
BRCA_miRNASeq <- test[,-1]
```

```{r}
# str(pw_miRNA)
# BRCA_miRNASeq <- t(BRCA_miRNASeq)
for(d in "BRCA"){
  print(d)
  for(o in omics[6]){
    print(o)
    dis_om <- sprintf("%s_%s", d, o)
    if(exists(dis_om)){
      assign(sprintf("%s_%s_pc1", d, o), 
             FirstPCPerPathway(
               om = o, 
               data = get(dis_om), 
               patients = get(sprintf("%s_patients", dis_om)), 
               pwlist = ifelse(
                 (o == "miRNA" | o == "miRNASeq"), 
                 list(pw_miRNA), 
                 list(kegg2)), 
               mseq = ifelse(
                 (o == "miRNASeq" | o == "mRNASeq"),
                 TRUE,
                 FALSE
               )
             )
      )
    }
  }
}
```
I need to set the rownames for CNV data to the gene names, maybe remove the second and third row?

```{r}
for(d in names(diseases)){
  objname <- sprintf("%s_CNV", d)
  obj <- get(objname)
  rownames(obj) <- obj[,1]
  assign(objname, obj[,c(-1, -2, -3)])
}
```
Same for the CNV patients dataframes:

```{r}
for(d in names(diseases)){
  objname <- sprintf("%s_CNV_patients", d)
  obj <- get(objname)
  assign(objname, obj[c(-1, -2, -3),])
}
```

Conclusion: for mRNA, CNV and methylation this already works. =) Now I need to find out what gene names are used for the sequencing data, and look up the gene targets of the microRNA's.

30-03-2017

The row names of the mRNAseq data are build up as follows: hgnc|entrez

```{r}
for(d in names(diseases)){
  print(d)
  objname <- sprintf("%s_mRNAseq", d)
  obj <- get(objname)
  rownames(obj) <- unlist(strsplit(rownames(obj), split = "[|]"))[seq(2, 2*nrow(obj), by = 2)]
  assign(objname, obj)
}
```

```{r}
o <- "mRNAseq"
for(d in names(diseases)){
  print(d)
  dis_om <- sprintf("%s_%s", d, o)
  if(exists(dis_om)){
    assign(sprintf("%s_%s_pc1", d, o), 
           FirstPCPerPathway(data = get(dis_om), patients = get(sprintf("%s_patients", dis_om)), mseq = TRUE)
    )
  }
}
```

Now I need to look up gene targets for miRNA's.

04-04-2017

Seeing if this method now also works for miRNA:

```{r}
test_GBM_miRNA <- FirstPCPerPathway(om = "miRNA", data = GBM_miRNA, patients = GBM_miRNA_patients, pwlist = pw_miRNA)
```

Okay, it works. It does look like for some pathways there are no data.

```{r}

for(d in names(diseases)){
  print(d)
  for(o in c("miRNA"
             )){
    print(o)
    dis_om <- sprintf("%s_%s", d, o)
    if(exists(dis_om)){
      assign(sprintf("%s_%s_pc1", d, o), 
             FirstPCPerPathway(om = o, data = get(dis_om), patients = get(sprintf("%s_patients", dis_om)), pwlist = pw_miRNA)
      )
    }
    else{print("doesn't exist")}
  }
}
```

Next step: check also the "average per pathway"-function to see if it will work with miRNA data. And convert the miRNASeq data so that only the reads_per_million are kept.

03-05-2017

It looks like the data is not as processed as it should be, as if the workspace has been reverted to an older state (CNV has the extra columns again, which makes me think maybe the other processing steps also have been reverted...).

Today the server will be restarted, so just to be sure I will leave this for after the restart.